C++: Add PreprocBlock.qll library

geoffw0 opened the pull request:
The `PreprocBlock.qll` library offers a view of `#if`, `#elif`, `#else` and similar directives (and the code they contain) as a tree.  That is, there's a parent-child relationship between blocks (identified by their opening directive), as well as the ability to tell what include statements and macro definitions are contained within each block.

---

The `preprocblock.expected` test is a good place to start if the above isn't crystal clear.

This is actually an old library, but until now it's not been available in the codeql repo.

MathiasVP commented on cpp/ql/lib/change-notes/2024-01-30-preproc-block.md:
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
```
I think this should actually be categorized as `feature`: https://github.com/github/codeql/blob/main/docs/change-notes.md#library-pack-change-categories

geoffw0 replied:
Good point.  Updated.

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/headers/PreprocBlock.qll:
```diff
@@ -0,0 +1,165 @@
+  /**
+   * Gets the file this `PreprocessorBlock` is located in.
+   */
+  cached
```
I'm slightly worried about all these `cached` annotations on all predicates. Other than possibly `getPreprocLineFromIndex`, none of these predicates look very expensive

geoffw0 replied:
Removing the `cached` annotations appears to _improve_ performance on some artificial code, and makes no consistent difference in the `RedundantInclude.ql` and `UselessInclude.ql` queries.  I'm certainly happier without annotations littered around, so I've removed them as suggested.  :+1:

MathiasVP reviewed (commented):
A few comments. I'd like to see two things fixed before we merge this:
- I think we should strip away all the `cached` and only re-introduce them if we actually end up seeing expensive re-evaluation. 
- I think we should fix all the implicit `this` warnings

But other than that this LGTM!

geoffw0 commented:
Thanks for the comments.  I agree and I've addressed the issues with implicit this and the change note.  Still to remove the `cached` annotations, I want to do that somewhat carefully to make sure performance isn't impacted.

MathiasVP reviewed (approved):
LGTM! I agree a DCA run isn't needed since nothing actually imports this library yet
